qa/tasks/cephfs/test_client_recovery.py

```diff
@@ -567,3 +567,37 @@ def test_abort_conn(self):
         self.assert_session_state(gid, "open")
         time.sleep(session_timeout * 1.5)  # Long enough for MDS to consider session stale
         self.assert_session_state(gid, "stale")
+
+    def test_dont_mark_unresponsive_client_stale(self):
+        """
+        Test that an unresponsive client holding caps not required by any
+        other client is evicted directly, without being marked as stale even
+        after session_timeout, at session_autoclose.
+        """
+        if not isinstance(self.mount_a, FuseMount):
+            self.skipTest("Require FUSE client to handle signal STOP/CONT")
+
+        # XXX: To conduct this test we need at least two clients since a
+        # single client is never evcited by MDS.
+        SESSION_TIMEOUT = 30
+        SESSION_AUTOCLOSE = 50
+        mount_a_gid = self.mount_a.get_global_id()
+        mount_a_pid = self.mount_a.client_pid
+        self.fs.set_var('session_timeout', SESSION_TIMEOUT)
+        self.fs.set_var('session_autoclose', SESSION_AUTOCLOSE)
+        self.assert_session_count(2, self.fs.mds_asok(['session', 'ls']))
+
+        # test that client holding cap not required by any other client is not
+        # marked stale when it becomes unresponsive.
+        self.mount_a.run_shell(['mkdir', 'dir'])
+        self.mount_a.send_signal('sigstop')
+        time.sleep(SESSION_TIMEOUT + 2)
+        self.assert_session_state(mount_a_gid, "open")
+
+        # test that other clients have to wait to get the caps from
+        # unresponsive client until session_autoclose.
+        self.mount_b.run_shell(['stat', 'dir'])
+        time.sleep(SESSION_AUTOCLOSE - SESSION_TIMEOUT + 2)
```

Comment:
Looking at this test more carefully again: The sleep is unnecessary. You can see from the log that the `stat` returns immediately, indicating that the conflicting caps were immediately reoslved and mount_a was evicted:

````
2019-07-13T16:57:54.484 INFO:tasks.cephfs.filesystem:_json_asok output: [
....
2019-07-13T16:57:54.484 INFO:teuthology.orchestra.run.smithi097:Running:
2019-07-13T16:57:54.484 INFO:teuthology.orchestra.run.smithi097:> cd /home/ubuntu/cephtest/mnt.1 && sudo stat dir
2019-07-13T16:57:57.595 INFO:teuthology.orchestra.run.smithi097.stdout:  File: ‘dir’
2019-07-13T16:57:57.595 INFO:teuthology.orchestra.run.smithi097.stdout:  Size: 0                Blocks: 1          IO Block: 4096   directory
2019-07-13T16:57:57.596 INFO:teuthology.orchestra.run.smithi097.stdout:Device: 2bh/43d  Inode: 1099511627776  Links: 2
2019-07-13T16:57:57.596 INFO:teuthology.orchestra.run.smithi097.stdout:Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
2019-07-13T16:57:57.596 INFO:teuthology.orchestra.run.smithi097.stdout:Context: system_u:object_r:fusefs_t:s0
2019-07-13T16:57:57.596 INFO:teuthology.orchestra.run.smithi097.stdout:Access: 2019-07-13 16:57:21.605488678 +0000
2019-07-13T16:57:57.596 INFO:teuthology.orchestra.run.smithi097.stdout:Modify: 2019-07-13 16:57:21.605488678 +0000
2019-07-13T16:57:57.596 INFO:teuthology.orchestra.run.smithi097.stdout:Change: 2019-07-13 16:57:21.605488678 +0000
2019-07-13T16:57:57.597 INFO:teuthology.orchestra.run.smithi097.stdout: Birth: -
````

from: /ceph/teuthology-archive/rishabh-2019-07-13_16:28:40-fs-wip-rishabh-defer-stale-clients-eviction-distro-basic-smithi/4117539/teuthology.log

from the MDS:

````
2019-07-13T16:57:54.483+0000 7f1ad137f700  1 mds.b asok_command: session ls (starting...)
2019-07-13T16:57:54.483+0000 7f1ad137f700  1 mds.b asok_command: session ls (complete)
2019-07-13T16:57:54.587+0000 7f1ad037d700  1 -- [v2:172.21.15.47:6832/2682686542,v1:172.21.15.47:6833/2682686542] <== client.7161 172.21.15.97:0/3128592429 10 ==== client_request(client.7161:3 getattr pAsLsXsFs #0x1 2019-07-13T16:57:54.585142+0000 caller_uid=1000, caller_gid=1206{6,1000,1206,10102,}) v4 ==== 168+0+0 (crc 0 0 0) 0x5573bec4adc0 con 0x5573bec51400
2019-07-13T16:57:54.587+0000 7f1ad037d700  4 mds.0.server handle_client_request client_request(client.7161:3 getattr pAsLsXsFs #0x1 2019-07-13T16:57:54.585142+0000 caller_uid=1000, caller_gid=1206{6,1000,1206,10102,}) v4
2019-07-13T16:57:54.587+0000 7f1ad037d700 20 mds.0.58 get_session have 0x5573bec50800 client.7161 172.21.15.97:0/3128592429 state open
...
2019-07-13T16:57:57.023+0000 7f1acdb78700 10 mds.0.server find_idle_sessions. last cleared laggy state 1251.48s ago
2019-07-13T16:57:57.023+0000 7f1acdb78700 10 mds.0.server new stale session client.5705 172.21.15.97:0/1713377504 last renewed caps 35.4159s ago
2019-07-13T16:57:57.023+0000 7f1acdb78700 20 mds.0.server laggiest active session is client.7161 172.21.15.97:0/3128592429 and renewed caps recently (8.77798s ago)
2019-07-13T16:57:57.023+0000 7f1acdb78700 10 mds.0.locker revoke_stale_caps for client.5705
2019-07-13T16:57:57.023+0000 7f1acdb78700 20 mds.0.server oldest stale session is client.5705 172.21.15.97:0/1713377504 and recently renewed caps 43.0269s ago
2019-07-13T16:57:57.023+0000 7f1acdb78700  0 log_channel(cluster) log [WRN] : evicting unresponsive client smithi097:0 (5705), after 43.0269 seconds
2019-07-13T16:57:57.023+0000 7f1acdb78700 10 mds.0.server autoclosing stale session client.5705 172.21.15.97:0/1713377504 last renewed caps 43.0269s ago
````

From: /ceph/teuthology-archive/rishabh-2019-07-13_16:28:40-fs-wip-rishabh-defer-stale-clients-eviction-distro-basic-smithi/4117539/remote/smithi047/log/ceph-mds.b.log.gz

You should next verify that we have not exceeded SESSION_AUTOCLOSE. We want the reason for the eviction to be the conflicting caps.

Reply:
> Looking at this test more carefully again: The sleep is unnecessary. You can see from the log that the stat returns immediately, indicating that the conflicting caps were immediately reoslved and mount_a was evicted

Ok. I've made the required change. This is from stdout when the test was run locally -
```
2019-07-19 13:04:48,757.757 INFO:__main__:Running ['stat', 'dir']
2019-07-19 13:04:53,238.238 INFO:__main__:Running ['./bin/ceph', 'daemon', 'mds.a', 'session', 'ls']
```

...and this from the teuthology log -

```
2019-07-19T06:00:20.120 INFO:teuthology.orchestra.run.smithi019:Running:
2019-07-19T06:00:20.120 INFO:teuthology.orchestra.run.smithi019:> cd /home/ubuntu/cephtest/mnt.1 && sudo stat dir
2019-07-19T06:00:22.993 INFO:teuthology.orchestra.run.smithi019.stdout:  File: ‘dir’
2019-07-19T06:00:22.993 INFO:teuthology.orchestra.run.smithi019.stdout:  Size: 0                Blocks: 1          IO Block: 4096   directory
2019-07-19T06:00:22.994 INFO:teuthology.orchestra.run.smithi019.stdout:Device: 2dh/45d  Inode: 1099511627776  Links: 2
2019-07-19T06:00:22.994 INFO:teuthology.orchestra.run.smithi019.stdout:Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
2019-07-19T06:00:22.994 INFO:teuthology.orchestra.run.smithi019.stdout:Context: system_u:object_r:fusefs_t:s0
2019-07-19T06:00:22.994 INFO:teuthology.orchestra.run.smithi019.stdout:Access: 2019-07-19 05:59:47.595784536 +0000
2019-07-19T06:00:22.995 INFO:teuthology.orchestra.run.smithi019.stdout:Modify: 2019-07-19 05:59:47.595784536 +0000
2019-07-19T06:00:22.995 INFO:teuthology.orchestra.run.smithi019.stdout:Change: 2019-07-19 05:59:47.595784536 +0000
2019-07-19T06:00:22.995 INFO:teuthology.orchestra.run.smithi019.stdout: Birth: -
```

It took much lesser time on teuthology than it takes locally. I hope that is not an issue.

> You should next verify that we have not exceeded SESSION_AUTOCLOSE. We want the reason for the eviction to be the conflicting caps.

Ok. Adding `self.assertLess(time.time(), time_at_beg + SESSION_AUTOCLOSE)` (where `time_at_beg` is the `time.time()` at the beginning of test) at end of the test.